frr-tools/status/controller/bgpsessionstate_controller.go

```diff
@@ -0,0 +1,283 @@
+	"github.com/metallb/frr-k8s/internal/frr"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+)
+
+const (
+	nodeLabel       = "frrk8s.metallb.io/node"
+	peerLabel       = "frrk8s.metallb.io/peer"
+	vrfLabel        = "frrk8s.metallb.io/vrf"
+	noBFDConfigured = "N/A"
+)
+
+type BGPPeersFetcher func() (map[string][]*frr.Neighbor, error)
+
+// BGPSessionStateReconciler reconciles a BGPSessionState object.
+type BGPSessionStateReconciler struct {
+	client.Client
+	BGPPeersFetcher
+	Logger       log.Logger
+	NodeName     string
+	Namespace    string
+	DaemonPod    *corev1.Pod
+	ResyncPeriod time.Duration
+}
+
+type peersStatuses map[string]*frrk8sv1beta1.BGPSessionState
+
+func (ps peersStatuses) add(s frrk8sv1beta1.BGPSessionState) {
+	ps[peerFor(s)] = s.DeepCopy()
+}
+
+func (ps peersStatuses) hasPeerFor(s frrk8sv1beta1.BGPSessionState) bool {
+	_, ok := ps[peerFor(s)]
+	return ok
+}
+
+func (ps peersStatuses) statusFor(peer string) *frrk8sv1beta1.BGPSessionState {
+	return ps[labelFormatForNeighbor(peer)]
+}
+
+func (ps peersStatuses) remove(peer string) {
+	delete(ps, labelFormatForNeighbor(peer))
+}
+
+// +kubebuilder:rbac:groups=frrk8s.metallb.io,resources=bgpsessionstates,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=frrk8s.metallb.io,resources=bgpsessionstates/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=frrk8s.metallb.io,resources=frrnodestates,verbs=get;list;watch
+// +kubebuilder:rbac:groups=frrk8s.metallb.io,resources=frrnodestates/status,verbs=get
+// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch
+
+func (r *BGPSessionStateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", "BGPSessionState", "start reconcile", req.NamespacedName.String())
+	defer level.Info(r.Logger).Log("controller", "BGPSessionState", "end reconcile", req.NamespacedName.String())
+
+	l := frrk8sv1beta1.BGPSessionStateList{}
+	err := r.Client.List(ctx, &l, client.MatchingLabels{nodeLabel: r.NodeName})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	existing, duplicates := peersStatusPerVRF(l.Items)
+	for _, s := range duplicates {
+		s := s
+		err := r.Client.Delete(ctx, &s)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+	}
+
+	neighbors, err := r.BGPPeersFetcher()
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	neighbors = renameDefaultVRF(neighbors)
+
+	toApply := map[string]peersStatuses{}
+	for vrf, neighs := range neighbors {
```

Comment:
neighs and  neighbors here are a bit hard to follow

Reply:
sorry if I insist :sweat_smile: but can you wrap this loop in a function which returns values that are easier to follow?

For example, below we range over existing because those are status belonging to non existing neighbors (I guess it's because the real "existing ones" are purged by that remove.

It'd be easier fi the function returned a "leftovers" value and a plain toApply slice (as opposed to map, as I don't think we use the vrf field anywhere.

Reply:
done